lib/db/schema/patch-084-085.sql

```diff
@@ -0,0 +1,204 @@
+--
+-- This migration updates the device-commands-related stored procedures
+-- to correctly use the (uid, deviceId) index when joining onto the
+-- `deviceCommands` table, fixing a performance issue.
+--
+
+SET NAMES utf8mb4 COLLATE utf8mb4_bin;
+
+CALL assertPatchLevel('84');
+
+CREATE PROCEDURE `accountDevices_15` (
+  IN `uidArg` BINARY(16)
+)
+BEGIN
+  SELECT
+    d.uid,
+    d.id,
+    d.sessionTokenId,
+    d.nameUtf8 AS name,
+    d.type,
+    d.createdAt,
+    d.callbackURL,
+    d.callbackPublicKey,
+    d.callbackAuthKey,
+    d.callbackIsExpired,
+    s.uaBrowser,
+    s.uaBrowserVersion,
+    s.uaOS,
+    s.uaOSVersion,
+    s.uaDeviceType,
+    s.uaFormFactor,
+    s.lastAccessTime,
+    ci.commandName,
+    dc.commandData
+  FROM devices AS d
+  INNER JOIN sessionTokens AS s
+    ON d.sessionTokenId = s.tokenId
+  LEFT JOIN (
+    deviceCommands AS dc
+    INNER JOIN deviceCommandIdentifiers AS ci
+      ON ci.commandId = dc.commandId
+  ) ON (dc.uid = d.uid AND dc.deviceId = d.id)
+  WHERE d.uid = uidArg
+  -- For easy flattening, ensure rows are ordered by device id.
+  ORDER BY 1, 2;
+END;
+
+CREATE PROCEDURE `device_2` (
+  IN `uidArg` BINARY(16),
+  IN `idArg` BINARY(16)
+)
+BEGIN
+  SELECT
+    d.uid,
+    d.id,
+    d.sessionTokenId,
+    d.nameUtf8 AS name,
+    d.type,
+    d.createdAt,
+    d.callbackURL,
+    d.callbackPublicKey,
+    d.callbackAuthKey,
+    d.callbackIsExpired,
+    s.uaBrowser,
+    s.uaBrowserVersion,
+    s.uaOS,
+    s.uaOSVersion,
+    s.uaDeviceType,
+    s.uaFormFactor,
+    s.lastAccessTime,
+    ci.commandName,
+    dc.commandData
+  FROM devices AS d
+  INNER JOIN sessionTokens AS s
+    ON d.sessionTokenId = s.tokenId
+  LEFT JOIN (
+    deviceCommands AS dc
+    INNER JOIN deviceCommandIdentifiers AS ci
+      ON ci.commandId = dc.commandId
+  ) ON (dc.uid = d.uid AND dc.deviceId = d.id)
+  WHERE d.uid = uidArg
+  AND d.id = idArg;
```

Comment:
⚠️ Full Index Scan

![image](https://user-images.githubusercontent.com/128755/44464043-a0e21680-a5e7-11e8-9fb8-6be9f944ef7d.png)

Reply:
Interesting.  MySQL appears to be choosing this query plan because it knows that the `deviceCommandIdentifiers` table is very small, so the full index scan on `ci` is cheap.  If it add a couple more rows to the `deviceCommandIdentifiers` table then it changes its plan to look up in `dc` before `ci`, which is what we'd expect.